### PR TITLE
rpma: add librpma_apm_* and librpma_gpspm_* engines

### PR DESCRIPTION
**Pull request by janekmi**
The Remote Persistent Memory Access (RPMA) Library is a C library created to simplify accessing persistent memory on remote hosts over Remote Direct Memory Access (RDMA).

The librpma_apm_client and librpma_apm_server is a pair of engines which allows benchmarking persistent writes achieved via
the Appliance Persistency Method (APM; natively supported by the librpma library) and regular reads (a part of the RDMA standard).

The librpma_gpspm_client and librpma_gpspm_server is a pair of engines which allows benchmarking persistent writes achieved via the General Purpose Persistency Method (GPSPM; build on top of the librpma API).

The librpma library is available here: https://github.com/pmem/rpma along with the set of scripts using the newly introduced engines to construct miscellaneous benchmarking scenarios: https://github.com/pmem/rpma/tree/master/tools/perf

The full history of the development of the librpma fio engines is available at: https://github.com/pmem/fio/tree/rpma

Co-Authored-By: Lukasz Dorau <lukasz.dorau@intel.com>
Co-Authored-By: Tomasz Gromadzki <tomasz.gromadzki@intel.com>
Co-Authored-By: Jan Michalski <jan.m.michalski@intel.com>
Co-Authored-By: Oksana Salyk <oksana.salyk@intel.com>

---
This change is [Reviewable](https://reviewable.io/reviews/axboe/fio/1186)

### CODE REVIEW

File: ci/travis-install-librpma.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash -e
+
+# 11.02.2021 Merge pull request #866 from ldorau/rpma-mmap-memory-for-rpma_mr_reg-in-rpma_flush_apm_new
+LIBRPMA_VERSION=fbac593917e98f3f26abf14f4fad5a832b330f5c
+
+WORKDIR=$(pwd)
+
+# install librpma
+git clone https://github.com/pmem/rpma.git
```
Comment (sitsofe) [inline ci/travis-install-librpma.sh:9]: Maybe grab the tarball like you do below? I guess technically I prefer to see curl used even if it needs more parameters :-)

File: ci/travis-install.sh [hunk 2]
```diff
@@ -43,6 +43,16 @@ case "$TRAVIS_OS_NAME" in
...
 	sudo apt-get install --no-install-recommends -qq -y "${pkgs[@]}"
+	# librpma is supported on the amd64 (x86_64) architecture for now
+	if [[ $CI_TARGET_ARCH == "amd64" ]]; then
+		# install libprotobuf-c-dev required by librpma_gpspm
+		sudo apt-get install --no-install-recommends -qq -y libprotobuf-c-dev
+		# PMDK libraries have to be installed, because
+		# libpmem is a dependency of the librpma fio engine
+		ci/travis-install-pmdk.sh
+		# install librpma from sources from GitHub
+		ci/travis-install-librpma.sh
```
Comment (sitsofe) [inline ci/travis-install.sh:54]: I wonder if these could just be functions?

File: configure [hunk 3]
```diff
@@ -920,6 +920,49 @@ if test "$disable_rdma" != "yes" && compile_prog "" "-lrdmacm" "rdma"; then
...
+#include <librpma.h>
+int main(int argc, char **argv)
+{
+  enum rpma_conn_event event = RPMA_CONN_REJECTED;
+  (void) event; /* unused */
+  rpma_log_set_threshold(RPMA_LOG_THRESHOLD, RPMA_LOG_LEVEL_INFO);
+  return 0;
+}
+EOF
+if test "$disable_rdma" != "yes" && compile_prog "" "-lrpma" "rpma"; then
```
Comment (sitsofe) [inline configure:939]: Where is this set? Do you need to add a flag further up? Further, we can probably do this all with `$librpma` by inverting the test to `if test "$librpma" != "no" ;` etc.

File: configure [hunk 4]
```diff
@@ -920,6 +920,49 @@ if test "$disable_rdma" != "yes" && compile_prog "" "-lrdmacm" "rdma"; then
...
+    librpma="yes"
+fi
+print_config "librpma" "$librpma"
+
+##########################################
+# libprotobuf-c probe
+if test "$libprotobuf_c" != "yes" ; then
+  libprotobuf_c="no"
+fi
```
Comment (sitsofe) [inline configure:948]: As above

File: engines/librpma_apm.c [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ */
+
+#include "librpma_fio.h"
+
+/* client side implementation */
+
+static inline int client_io_flush(struct thread_data *td,
+		struct io_u *first_io_u, struct io_u *last_io_u,
```
Comment (sitsofe) [inline engines/librpma_apm.c:22]: I think the house style is tending towards line up with bracket

File: engines/librpma_apm.c [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+	struct librpma_fio_client_data *ccd;
+	unsigned int sq_size;
+	uint32_t cq_size;
+	struct rpma_conn_cfg *cfg = NULL;
+	struct rpma_peer_cfg *pcfg = NULL;
+	int ret;
+
+	/* not supported readwrite = trim / randtrim / trimwrite */
+	if (td_trim(td)) {
+		log_err("Not supported mode.\n");
```
Comment (sitsofe) [inline engines/librpma_apm.c:39]: I'd set `td_verror` like in engines/ime.c. Maybe return `-EINVAL`?

File: engines/librpma_apm.c [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+	if (librpma_fio_client_init(td, cfg))
+		goto err_cfg_delete;
+
+	ccd = td->io_ops_data;
+
+	if (ccd->server_mr_flush_type == RPMA_FLUSH_TYPE_PERSISTENT) {
+		if (!ccd->ws->direct_write_to_pmem) {
+			if (td->thread_number == 1)
+				log_err(
+					"Fio librpma engine will not work until the Direct Write to PMem on the server side is possible (direct_write_to_pmem)\n");
```
Comment (sitsofe) [inline engines/librpma_apm.c:110]: Could you split thi line?

```
log_err("initial text"
             "more text"
             "etc\n")

File: engines/librpma_apm.c [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+				"rpma_conn_apply_remote_peer_cfg");
+			goto err_cleanup_common;
+		}
+
+		(void) rpma_peer_cfg_delete(&pcfg);
+	} else if (td->thread_number == 1) {
+		/* XXX log_info mixes with the JSON output */
+		log_err(
+			"Note: Direct Write to PMem is not supported by default nor required if you use DRAM instead of PMem on the server side (direct_write_to_pmem).\n"
+			"Remember that flushing to DRAM does not make your data persistent and may be used only for experimental purposes.\n");
```
Comment (sitsofe) [inline engines/librpma_apm.c:137]: Can you wrap this as above.

File: engines/librpma_apm.c [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+		if ((ret = rpma_conn_apply_remote_peer_cfg(ccd->conn, pcfg))) {
+			librpma_td_verror(td, ret,
+				"rpma_conn_apply_remote_peer_cfg");
+			goto err_cleanup_common;
+		}
+
+		(void) rpma_peer_cfg_delete(&pcfg);
+	} else if (td->thread_number == 1) {
+		/* XXX log_info mixes with the JSON output */
```
Comment (sitsofe) [inline engines/librpma_apm.c:134]: That's a more generic issue (see https://github.com/axboe/fio/issues/731 ) so I don't think you need to comment it here.

File: engines/librpma_apm.c [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+		log_err(
+			"Note: Direct Write to PMem is not supported by default nor required if you use DRAM instead of PMem on the server side (direct_write_to_pmem).\n"
+			"Remember that flushing to DRAM does not make your data persistent and may be used only for experimental purposes.\n");
+	}
+
+	if ((ret = rpma_conn_cfg_delete(&cfg))) {
+		librpma_td_verror(td, ret, "rpma_conn_cfg_delete");
+		/* non fatal error - continue */
```
Comment (sitsofe) [inline engines/librpma_apm.c:142]: Maybe move the comment one line up.

File: engines/librpma_gpspm_flush.pb-c.c [hunk 11]
```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2020, Intel Corporation
```
Comment (sitsofe) [inline engines/librpma_gpspm_flush.pb-c.c:2]: Technically 2021?

File: engines/librpma_fio.h [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+
+#include <librpma.h>
+
+/* servers' and clients' common */
+
+#define librpma_td_verror(td, err, func) \
+	td_vmsg((td), (err), rpma_err_2str(err), (func))
+
+/* ceil(a / b) = (a + b - 1) / b */
+#define LIBRPMA_FIO_CEIL(a, b) (((a) + (b) - 1) / (b))
```
Comment (sitsofe) [inline engines/librpma_fio.h:30]: What's wrong with`ceil()`? Too slow?

File: engines/librpma_fio.h [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+#include "../optgroup.h"
+
+#include <librpma.h>
+
+/* servers' and clients' common */
+
+#define librpma_td_verror(td, err, func) \
+	td_vmsg((td), (err), rpma_err_2str(err), (func))
+
+/* ceil(a / b) = (a + b - 1) / b */
```
Comment (sitsofe) [inline engines/librpma_fio.h:29]: I think you don't need this comment as it literally describes the macro!

File: engines/librpma_fio.h [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+#define LIBRPMA_FIO_CEIL(a, b) (((a) + (b) - 1) / (b))
+
+/* common option structure for server and client */
+struct librpma_fio_options_values {
+	/*
+	 * FIO considers .off1 == 0 absent so the first meaningful field has to
+	 * have padding ahead of it.
+	 */
+	void *pad;
+	/* server ip */
```
Comment (sitsofe) [inline engines/librpma_fio.h:39]: Maybe skip the obvious comments that just re-describe what the code says code (but keep the non-obvious ones).

File: engines/librpma_fio.h [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+		unsigned long long int len);
+
+/*
+ * RETURN VALUE
+ * - ( 1) - on success
+ * - ( 0) - skip
+ * - (-1) - on error
+ */
+typedef int (*librpma_fio_get_io_u_index_t)(struct rpma_completion *cmpl,
+		unsigned int *io_u_index);
```
Comment (sitsofe) [inline engines/librpma_fio.h:97]: @axboe how do you feel about this typedef?

File: engines/librpma_gpspm.c [hunk 16]
```diff
@@ -0,0 +1,755 @@
...
+
+	/*
+	 * not supported:
+	 * - readwrite = read / trim / randread / randtrim /
+	 *               / rw / randrw / trimwrite
+	 */
+	if (td_read(td) || td_trim(td)) {
+		log_err("Not supported mode.\n");
+		return -1;
+	}
```
Comment (sitsofe) [inline engines/librpma_gpspm.c:82]: Above suggestions in a different file apply here too:
- Skip non obvious comments
- Try and wrap to line up with the brackets
- Skip obvious comments but keep the non-obvious comments

File: fio.1 [hunk 17]
```diff
@@ -1958,6 +1958,12 @@ The hostname or IP address to use for TCP, UDP or RDMA-CM based I/O.
 If the job is a TCP listener or UDP reader, the hostname is not used
 and must be omitted unless it is a valid UDP multicast address.
 .TP
+.BI (librpma_*)serverip \fR=\fPstr
+The IP address to be used for RDMA-CM based I/O.
+.TP
+.BI (librpma_*_server)direct_write_to_pmem \fR=\fPbool
+Set to 1 only when Direct Write to PMem from the remote host is possible. Otherwise, set to 0.
+.TP
```
Comment (sitsofe) [inline fio.1:1966]: You'll need to document this in the HOWTO too.

**Review by sitsofe — CHANGES_REQUESTED** [review]
Phew, this one is big! Some general comments but nothing to severe. @axboe given the sheer number of files should these engines go into a subdirectory?

**Review by ldorau — COMMENTED** [review]
> Reviewed 20 of 20 files at r1.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @janekmi)

**Review by janekmi — COMMENTED** [review]
> Reviewed 1 of 20 files at r1.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @janekmi and @sitsofe)

___
*[ci/travis-install.sh, line 54 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WQ-r1-54:-MTdmROE5PUbhs5Ign34:b-wp6y3) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/ci/travis-install.sh#L54)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

I wonder if these could just be functions?
</blockquote></details>

I think it is handier to have them as scripts since they can be easily used to set up your own environment not only for setting up ci.
___
*[engines/librpma_apm.c, line 22 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WS-r1-22:-MTdo2E25mk-NMoGqfcz:b654zy7) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_apm.c#L22)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

I think the house style is tending towards line up with bracket
</blockquote></details>

Can you please write exactly how you imagine adjusting this very line, so I can use it as an example for adjusting others?
___
*[engines/librpma_apm.c, line 110 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WS-r1-110:-MTdopeKEmNBVvisbJwh:bgx27ub) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_apm.c#L110)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

Could you split thi line?

```
log_err("initial text"
             "more text"
             "etc\n")
</blockquote></details>

I like to have string literals in a single line so it is easier to find in the codebase. Otherwise, you have to guess if and where it may have been got split.
___
*[engines/librpma_apm.c, line 134 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WS-r1-134:-MTdpUOfBFH7srb10NFF:ba2epup) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_apm.c#L134)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

That's a more generic issue (see https://github.com/axboe/fio/issues/731 ) so I don't think you need to comment it here.
</blockquote></details>

Having an issue is great. I can refer to it here. But I would argue you won't look for issues when you browse through the source code of an engine and you find out something nonintuitive is going on. This is where this comment proves helpful.
___
*[engines/librpma_apm.c, line 137 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WS-r1-137:-MTdpwkx7JezrQipEn3q:bgx27ub) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_apm.c#L137)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

Can you wrap this as above.
</blockquote></details>

I like to have string literals in a single line so it is easier to find in the codebase. Otherwise, you have to guess if and where it may have been got split.
___
*[engines/librpma_apm.c, line 142 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WS-r1-142:-MTdq5lq7Tw69BB2CU9V:b17erzp) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_apm.c#L142)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

Maybe move the comment one line up.
</blockquote></details>

This comment is here instead of some kind of an exit/error one would expect. Having it above the logging call would suggest it describes the logging call which it does not.


<!-- Sent from Reviewable.io -->

**Review by janekmi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @janekmi and @sitsofe)

___
*[configure, line 939 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WR-r1-939:-MTdsncC5QsM_LiKBN66:bfu95i3) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/configure#L939)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

Where is this set? Do you need to add a flag further up? Further, we can probably do this all with `$librpma` by inverting the test to `if test "$librpma" != "no" ;` etc.
</blockquote></details>

`$disable_rdma` is set here: https://github.com/janekmi/fio/blob/add-librpma-engines/configure#L205
___
*[configure, line 948 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WR-r1-948:-MTdtJJY8u8P2I5xEhky:b-vh8szk) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/configure#L948)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

As above
</blockquote></details>

I do not follow how you would prefer to change these conditions. Can you write exactly your idea, please?

**Review by ldorau — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, and @sitsofe)

___
*[ci/travis-install-librpma.sh, line 9 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WO-r1-9:-MTdvOqz1vzIrmhpjScQ:b-t37jav) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/ci/travis-install-librpma.sh#L9)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

Maybe grab the tarball like you do below? I guess technically I prefer to see curl used even if it needs more parameters :-)
</blockquote></details>

There is no tarball containing the required commit yet.

**Review by janekmi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, and @sitsofe)

___
*[engines/librpma_fio.h, line 29 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WU-r1-29:-MTe2p9z6mtHvw5nAQ4q:b-xfhz6p) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_fio.h#L29)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

I think you don't need this comment as it literally describes the macro!
</blockquote></details>

Not exactly. It also describes the role of the parameters.
___
*[engines/librpma_gpspm_flush.pb-c.c, line 2 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WW-r1-2:-MTdtswb5ioSdxUfEAku:b6tdfkn) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_gpspm_flush.pb-c.c#L2)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

Technically 2021?
</blockquote></details>

Yeah... we have started the development last year and this particular file hasn't changed since 2020.

**Review by janekmi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, and @sitsofe)

___
*[engines/librpma_gpspm.c, line 82 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WV-r1-82:-MTe4SlL0GjTkqSn0yTd:bds359i) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_gpspm.c#L82)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

Above suggestions in a different file apply here too:
- Skip non obvious comments
- Try and wrap to line up with the brackets
- Skip obvious comments but keep the non-obvious comments
</blockquote></details>

1. I have browsed the code one more time. No obvious comments in this particular file have been found. Please, point them out exactly so we can discuss them one-by-one.
2. I have asked for an example in one of your previous comments. I will be happy to do this once we agree on what is to be done.
3. :-)

**Review by janekmi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, and @sitsofe)

___
*[ci/travis-install-librpma.sh, line 9 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WO-r1-9:-MTe5TJQ5z0cfGRHFpUR:b-housix) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/ci/travis-install-librpma.sh#L9)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

There is no tarball containing the required commit yet.
</blockquote></details>

I guess we can still download a zip as follows: https://github.com/pmem/rpma/archive/fbac593917e98f3f26abf14f4fad5a832b330f5c.zip

**Review by janekmi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: 17 of 20 files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, @ldorau, and @sitsofe)

___
*[engines/librpma_apm.c, line 39 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WS-r1-39:-MTe71KkC6KgC3kea5GT:b-896fix) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_apm.c#L39)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

I'd set `td_verror` like in engines/ime.c. Maybe return `-EINVAL`?
</blockquote></details>

Done.
___
*[engines/librpma_fio.h, line 39 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WU-r1-39:-MTe75dw9gPKfjapEEF5:b-896fix) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_fio.h#L39)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

Maybe skip the obvious comments that just re-describe what the code says code (but keep the non-obvious ones).
</blockquote></details>

Done.

**Review by ldorau — COMMENTED** [review]
> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, and @sitsofe)

**Review by ldorau — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, and @sitsofe)

___
*[ci/travis-install-librpma.sh, line 9 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WO-r1-9:-MTeLdeyAzH07J7WZ_zR:b-bnj460) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/ci/travis-install-librpma.sh#L9)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

I guess we can still download a zip as follows: https://github.com/pmem/rpma/archive/fbac593917e98f3f26abf14f4fad5a832b330f5c.zip
</blockquote></details>

Like that? https://github.com/pmem/fio/pull/209/files

**Review by janekmi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, and @sitsofe)

___
*[engines/librpma_fio.h, line 30 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WU-r1-30:-MTeSvIs2qoYj6x8ChjW:bjytkln) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/engines/librpma_fio.h#L30)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

What's wrong with`ceil()`? Too slow?
</blockquote></details>

I don't like casting integers to doubles and back to integers for something which can be done easily with integers.

**Review by janekmi — COMMENTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, and @sitsofe)

___
*[ci/travis-install-librpma.sh, line 9 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecnTZXofGuGR0WO-r1-9:-MTeVR9T1GKCqCT9pAAA:b-896fix) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/ci/travis-install-librpma.sh#L9)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

Like that? https://github.com/pmem/fio/pull/209/files
</blockquote></details>

Done.

**Review by ldorau — COMMENTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, and @sitsofe)

**Review by ldorau — COMMENTED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe, @janekmi, and @sitsofe)

**Review by janekmi — COMMENTED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe and @sitsofe)

___
*[fio.1, line 1966 at r1](https://reviewable.io/reviews/axboe/fio/1186#-MTHsecosg3IPqXsuzz1-r1-1966:-MTeX4reAE7ZJRN70uib:b-896fix) ([raw file](https://github.com/axboe/fio/blob/061b6f34439257bb3b656c21ecf838d6180a51cd/fio.1#L1966)):*
<details><summary><i>Previously, sitsofe (Sitsofe Wheeler) wrote…</i></summary><blockquote>

You'll need to document this in the HOWTO too.
</blockquote></details>

Done.

**Review by janekmi — COMMENTED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe and @sitsofe)

**Review by janekmi — COMMENTED** [review]
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe and @sitsofe)

**Review by ldorau — COMMENTED** [review]
> Reviewed 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe and @sitsofe)

**Review by ldorau — COMMENTED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/axboe/fio/1186)* status: all files reviewed, 17 unresolved discussions (waiting on @axboe and @sitsofe)

**Pull request merged**